blog/2024-05-21-blog-status-of-lbse-in-webkit.md

```diff
@@ -0,0 +1,149 @@
+
+## Resource invalidation handling
+
+After introducing the first SVG resource (`RenderSVGResourceClipper`) we noticed some issues with handling invalidations (for example adding to clip-path contents) for it. In the legacy engine, invalidations have been handled through layouting. This caused various problems, for one it could cause calling the setNeedsLayout method from within layout and the invalidation chain depended on the DOM order.
+
+In **November 2023** an <a href="https://commits.webkit.org/271017@main">implementation</a> landed that avoided using layout for resource invalidation. Instead, on dynamic updates the style system is used to determine the appropriate action:
+
+- For non resources, changes that cause renderer geometry changes, like changing the x value of a &lt;rect&gt; element still require a relayout. For visual changes not affecting geometry, like changing fill color on a &lt;rect&gt;, a repaint action is enough.
+- For resources, the resource is invalidated/updated with the change and any of its clients are repainted using the new resource.
+
+## Support for masks
+
+With improved support for SVG resource invalidation, in **late November 2023** we were ready to upstream support for the next SVG resource, <a href="https://commits.webkit.org/271153@main">RenderSVGResourceMasker</a>.
+
+Like the support for clip-path, it started out without caching image buffers and makes use of Porter-Duff DestinationIn/SourceOver compositing to efficiently render the mask.
+
+Mask content invalidations/changes were supported out of the box since we had improved resource invalidation handling (see above).
+
+## Support for gradients
+
+In early **January 2024** support for SVG gradients was <a href="https://commits.webkit.org/272653@main">upstreamed</a>. This is a kind of SVG resource that is a bit different to the previously implemented clipping paths/masks because it is a <a href="">paint server</a>, so a helper class for that called `SVGPaintServerHandling` and a base class `RenderSVGResourcePaintServer` were introduced.
+
+## Support for patterns
+
+End of **January 2024** support for SVG patterns was  <a href="https://commits.webkit.org/273757@main">upstreamed</a>. In the first implementation, no image buffer caching was implemented to keep things clean and simple. The implementation is different from the legacy implementation because the pattern contents are being rendered through the pattern content layers (See <a href="https://github.com/WebKit/WebKit/blob/main/Source/WebCore/rendering/RenderLayer.cpp#L2846">`RenderLayer::paintSVGResourceLayer`</a>). To make this work, `RenderSVGResourcePattern` has to setup the graphics context matrix correctly before calling `paintSVGResourceLayer`.
+
+## Support for markers
+
+Also in end of **January 2024**, we tackled the last remaining SVG resource on out TODO list, namely <a href="https://commits.webkit.org/273820@main">markers</a>.
```

Comment:
The next-to-last resource... This should be rephrased.

Reply:
Done.